jupyterhub/app.py

```diff
@@ -2014,14 +2014,25 @@ async def init_groups(self):
 
     async def init_role_creation(self):
         """Load default and predefined roles into the database"""
-        self.log.debug('Loading default roles to database')
+        self.log.debug('Loading roles into database')
         default_roles = roles.get_default_roles()
         config_role_names = [r['name'] for r in self.load_roles]
 
-        init_roles = default_roles
+        default_roles_dict = {role["name"]: role for role in default_roles}
+        init_roles = []
         roles_with_new_permissions = []
         for role_spec in self.load_roles:
             role_name = role_spec['name']
+            if role_name in default_roles_dict:
```

Comment:
Reworked how default roles were overridden - previously, they were loaded twice - once with default spec, then again with any overrides. This merges the spec so that there aren't two logged changes to the role's permissions when there's no actual change in config (change from db to default; change from default to config, which matched db).